### PR TITLE
[Prototype] Retries for Storage

### PR DESCRIPTION
**Pull request by frankyn**
Prototype based on BigQuery retry implementation in Python: https://github.com/googleapis/python-bigquery/blob/master/google/cloud/bigquery/retry.py

Work towards resolving: #108

Remaining:

- [ ] Idempotent operations are retried *only*
- [ ] Reset Connection Approval
- [ ] Unit Tests
- [ ] Review

### CODE REVIEW

File: google/cloud/storage/_helpers.py [hunk 1]
```diff
@@ -134,6 +136,14 @@ def _query_params(self):
             params["userProject"] = self.user_project
         return params
 
+
+    def _call_api(self, client, retry, **kwargs):
+        call = functools.partial(client._connection.api_request, **kwargs)
```
Comment (andrewsg) [inline google/cloud/storage/_helpers.py:141]: Is this because the interface for the retry function is not expected to pass in any arguments?

Comment (frankyn) [reply google/cloud/storage/_helpers.py:141]: Copied it over mainly:
At the moment no, I'm not making it configurable for the first pass and only introducing it to help support/unblock customer friction when using the library.

File: google/cloud/storage/retry.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+"""Retry Strategy logic used across google.cloud.storage requests."""
+
+# Should not be in here and only for prototyping
+import six
+import socket
+import requests
+import urllib3
+
+from google.api_core import exceptions
+from google.api_core import retry
```
Comment (tritone) [inline google/cloud/storage/retry.py:23]: Neat, so this is what we told customers to use as a workaround already right?

File: google/cloud/storage/retry.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    """Predicate for determining when to retry."""
+
+    if hasattr(exc, "errors"):
+        if len(exc.errors) == 0:
+            # Check for unstructured error returns, e.g. from GFE
+            return isinstance(exc, _UNSTRUCTURED_RETRYABLE_TYPES)
+        reason = exc.errors[0]["reason"]
+
+        return reason in _RETRYABLE_REASONS
+    else:
```
Comment (tritone) [inline google/cloud/storage/retry.py:49]: No need for `else` here right? And the nested if statements and uses of isinstance look pretty ugly, I'm sure there's a nicer way to do this.

File: google/cloud/storage/retry.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+def _should_retry(exc):
+    """Predicate for determining when to retry."""
+
+    if hasattr(exc, "errors"):
+        if len(exc.errors) == 0:
+            # Check for unstructured error returns, e.g. from GFE
+            return isinstance(exc, _UNSTRUCTURED_RETRYABLE_TYPES)
+        reason = exc.errors[0]["reason"]
+
+        return reason in _RETRYABLE_REASONS
```
Comment (tritone) [inline google/cloud/storage/retry.py:48]: Do you not have access to response codes here? Seems kind of a strange workaround to be doing string matching for a 503 for example.

File: google/cloud/storage/retry.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    else:
+        # Connection Reset
+        if isinstance(exc, requests.exceptions.ConnectionError):
+            if isinstance(exc.args[0], urllib3.exceptions.ProtocolError):
+                if isinstance(exc.args[0].args[1], ConnectionResetError):
+                    return True
+    return False
+
+_DEFAULT_RETRY = retry.Retry(predicate=_should_retry)
```
Comment (tritone) [inline google/cloud/storage/retry.py:57]: I would be curious to look at other examples of where people have written these predicates.

File: google/cloud/storage/retry.py (hunk 2 above)
Comment (tritone) [reply google/cloud/storage/retry.py:23]: Also, where are configs around number of attempts or timeouts? Are these values hard-coded into google.api_core.retry?

File: google/cloud/storage/_helpers.py [hunk 6]
```diff
@@ -134,6 +136,14 @@ def _query_params(self):
             params["userProject"] = self.user_project
         return params
 
+
+    def _call_api(self, client, retry, **kwargs):
```
Comment (tritone) [inline google/cloud/storage/_helpers.py:140]: Should retry be a kwarg? Do callers explicitly have to pass in None if they don't want a retry? Would be good to clarify here.

**Comment by simonvanderveldt** [comment]
Is there any chance of this being merged soon? We'd very much like to have this functionality available in the library, we currently have to wrap everything in retries ourselves.

**Comment by simonvanderveldt** [comment]
@busunkim96 Why is this closed? Is there a new PR?

**Comment by andrewsg** [comment]
I'll be working on a new one.

**Comment by andyhky** [comment]
Is there any update on this?

We recently encountered no retries for `google.cloud.storage.blob.Blob.reload` https://googleapis.dev/python/storage/latest/blobs.html?#google.cloud.storage.blob.Blob.reload and it would be nice to have this functionality.

**Comment by frankyn** [comment]
Closing this PR as @andrewsg is working on a new PR now. PR should be available soon. Thank you for your patience.

**Pull request closed without merging**